thinc/loss.py

```diff
@@ -58,142 +171,190 @@ def __init__(
         else:
             self._name_to_i = {}
 
-    def convert_truths(self, truths, guesses: Floats2d) -> Tuple[Floats2d, Floats2d]:
-        xp = get_array_module(guesses)
+    def __call__(
+        self, guesses: Floats2d, truths: Union[Sequence[int], Sequence[str]]
+    ) -> Tuple[Floats2d, float]:
+        target, mask = self.convert_truths(truths, guesses)
+        self._validate_input(guesses, target)
+        d_truth = self._get_grad(guesses, target, mask)
+        loss = self._get_loss(guesses, target, mask)
+        return (d_truth, loss)
+
+    def _check_ints1d(self, arr):
```

Comment:
We should add a type hint for `arr`. According to the one call to this function in this file, it should be `Categories1d`?

Reply:
I moved it out of the loss and then it can have its more true `arr: ArrayXd` signature.